### PR TITLE
Add support for EXCEPT

### PR DESCRIPTION
**Pull request by rschlussel-zz**
@anusudarsan

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementExceptAsUnion.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        private final SymbolAllocator symbolAllocator;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitExcept(ExceptNode node, RewriteContext<Void> rewriteContext)
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementExceptAsUnion.java]: can we move this to ImplementIntersectAsUnion (rename it) and re-use the methods there wherever possible?

**Comment by rschlussel-zz** [comment]
updated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java [hunk 2]
```diff
@@ -125,26 +129,52 @@ public PlanNode visitIntersect(IntersectNode node, RewriteContext<Void> rewriteC
...
+
+            // add a union over all the rewritten sources. The outputs of the union have the same name as the
+            // original except node
+            List<Symbol> outputs = node.getOutputSymbols();
+            UnionNode union = union(withMarkers, ImmutableList.copyOf(concat(outputs, markers)));
+
+            // add count aggregations and filter rows where count for the first source is >= 1 and all others are 0
+            List<Symbol> aggregationOutputs = allocateSymbols(markers.size(), "count", BIGINT);
+            AggregationNode aggregation = computeCounts(union, outputs, markers, aggregationOutputs);
+            FilterNode filterNode = addFilterForExcept(aggregation, aggregationOutputs);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:177]: is it possible to get the symbol list `aggregationOutputs` from `aggregation`? so then we need to pass only the `AggregationNode` to the method?

File: presto-product-tests/src/main/resources/sql-tests/testcases/set_operation/except.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
+-- database: presto; requires: com.facebook.presto.tests.ImmutableTpchTablesRequirements; tables: nation, workers; groups: set_operation;
+-- delimiter: |; ignoreOrder: true;
+--! name: except_union_intersect
+SELECT n_name FROM nation WHERE n_nationkey = 17 
+EXCEPT
+SELECT n_name FROM nation WHERE n_regionkey = 2
+UNION
```
Comment (anusudarsan) [inline presto-product-tests/src/main/resources/sql-tests/testcases/set_operation/except.sql]: maybe another test with UNION ALL, EXCEPT, INTERSECT all in one query?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java [hunk 4]
```diff
@@ -78,7 +81,7 @@
  * GROUP BY a) T2
  * WHERE foo_cnt >= 1 AND bar_cnt >= 1;
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:82]: update java docs to include EXCEPT

**Comment by anusudarsan** [comment]
lgtm % comments

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java (hunk 2 above)
Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:177]: You can't get an ordered list, and I needed to know which one is the first one.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java (hunk 4 above)
Comment (rschlussel-zz) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAndExceptAsUnion.java:82]: good point.  I should also update the user docs.

**Comment by rschlussel-zz** [comment]
updated

**Comment by anusudarsan** [comment]
LGTM. thanks!

**Comment by kbajda** [comment]
Do you expect TPC-DS query 87 to start working after this PR?

**Comment by rschlussel-zz** [comment]
@KBP-TDC yes, i expect so.  I don't see anything else in that query that is unsupported.

**Pull request closed without merging**